src/python/WMCore/WorkQueue/WorkQueue.py

```diff
@@ -816,8 +817,29 @@ def closeWork(self, *workflows):
         self.backend.recordTaskActivity('workclosing', msg)
 
         return workflowsToClose
-
-    def performQueueCleanupActions(self, skipWMBS = False):
+    
+    def deleteCompletedWFElements(self):
+        """
+        deletes Workflow when workflow is in finished status
+        """
+        reqNames = self.backend.getWorkflowNames()
+        reqNamesInbox = self.backend.flowNames(inboxFlag = True)
+        reqNames = list(set(reqNames).union(set(reqNamesInbox)))
+        requestsInfo = self.reqmgrSvc.getRequestByNames(reqNames)
+        deleteRequests = []
+        for key, value in requestsInfo.items():
+            if ((value["RequestStatus"] == None) or \
```

Comment:
The problem is deleting WQ is parallel to state change. 
Workflow can be archived before WQ elements gets deleted. If that happens,WQ elements will never get deleted. Getting status of the workflow should include the archived workflow if we want to prevent that. But that will be too expensive. Getting workflow from WQ will limit the active workflows (~1000). I think that could be more efficient. (I need to check whether that call is ' POST' call though).